From e91b8d3f6fd69e0080bbfc891816f91c7bcf7a8f Mon Sep 17 00:00:00 2001
From: Miao Yu <r65093@freescale.com>
Date: Wed, 22 Apr 2009 12:40:24 +0800
Subject: [PATCH] ENGR00109080 MX35: Add TVIN power-saving mode

1. Support TVIN power-saving mode when system enter suspend/resume.
2. Revise SENS_FRM_HEIGHT and SENS_FRM_WIDTH register value,
(SENS_FRM_HEIGHT is equal to the sensor frame height - 1)
otherwise there will be a white dotted line at the top sometimes.

Signed-off-by: Miao Yu <r65093@freescale.com>
---
 drivers/media/video/mxc/capture/adv7180.c          |   31 ++++++++++++-------
 drivers/media/video/mxc/capture/mxc_v4l2_capture.c |   18 ++++++------
 drivers/mxc/ipu/ipu_common.c                       |   11 ++++++-
 3 files changed, 38 insertions(+), 22 deletions(-)

diff --git a/drivers/media/video/mxc/capture/adv7180.c b/drivers/media/video/mxc/capture/adv7180.c
index 42c8c28..07a68ec 100644
--- a/drivers/media/video/mxc/capture/adv7180.c
+++ b/drivers/media/video/mxc/capture/adv7180.c
@@ -117,24 +117,24 @@ static video_fmt_t video_fmts[] = {
 	{			/*! NTSC */
 	 .v4l2_id = V4L2_STD_NTSC,
 	 .name = "NTSC",
-	 .raw_width = 720,	/* SENS_FRM_WIDTH */
-	 .raw_height = 288,	/* SENS_FRM_HEIGHT */
-	 .active_width = 720,	/* ACT_FRM_WIDTH */
-	 .active_height = (480 / 2),	/* ACT_FRM_WIDTH */
+	 .raw_width = 720 - 1,	/* SENS_FRM_WIDTH */
+	 .raw_height = 288 - 1,	/* SENS_FRM_HEIGHT */
+	 .active_width = 720,	/* ACT_FRM_WIDTH plus 1 */
+	 .active_height = (480 / 2),	/* ACT_FRM_WIDTH plus 1 */
 	 },
 	{			/*! (B, G, H, I, N) PAL */
 	 .v4l2_id = V4L2_STD_PAL,
 	 .name = "PAL",
-	 .raw_width = 720,
-	 .raw_height = (576 / 2) + 24 * 2,
+	 .raw_width = 720 - 1,
+	 .raw_height = (576 / 2) + 24 * 2 - 1,
 	 .active_width = 720,
 	 .active_height = (576 / 2),
 	 },
 	{			/*! Unlocked standard */
 	 .v4l2_id = V4L2_STD_ALL,
 	 .name = "Autodetect",
-	 .raw_width = 720,
-	 .raw_height = (576 / 2) + 24 * 2,
+	 .raw_width = 720 - 1,
+	 .raw_height = (576 / 2) + 24 * 2 - 1,
 	 .active_width = 720,
 	 .active_height = (576 / 2),
 	 },
@@ -160,6 +160,7 @@ static DECLARE_MUTEX(mutex);
 #define ADV7180_MANUAL_WIN_CTL         0x3d	/* Manual Window Control */
 #define ADV7180_SD_SATURATION_CB       0xe3	/* SD Saturation Cb */
 #define ADV7180_SD_SATURATION_CR       0xe4	/* SD Saturation Cr */
+#define ADV7180_PWR_MNG                0x0f     /* Power Management */
 
 /* supported controls */
 /* This hasn't been fully implemented yet.
@@ -331,12 +332,17 @@ static int ioctl_s_power(struct v4l2_int_device *s, int on)
 
 	dev_dbg(&adv7180_data.i2c_client->dev, "In adv7180:ioctl_s_power\n");
 
-	sensor->on = on;
-
-	if (on)
+	if (on && !sensor->on) {
 		gpio_sensor_active();
-	else
+		if (adv7180_write_reg(ADV7180_PWR_MNG, 0) != 0)
+			return -EIO;
+	} else if (!on && sensor->on) {
+		if (adv7180_write_reg(ADV7180_PWR_MNG, 0x24) != 0)
+			return -EIO;
 		gpio_sensor_inactive();
+	}
+
+	sensor->on = on;
 
 	return 0;
 }
@@ -859,6 +865,7 @@ static int adv7180_probe(struct i2c_client *client,
 	adv7180_data.pix.height = video_fmts[video_idx].raw_height;
 	adv7180_data.pix.pixelformat = V4L2_PIX_FMT_UYVY;  /* YUV422 */
 	adv7180_data.pix.priv = 1;  /* 1 is used to indicate TV in */
+	adv7180_data.on = true;
 
 	gpio_sensor_active();
 
diff --git a/drivers/media/video/mxc/capture/mxc_v4l2_capture.c b/drivers/media/video/mxc/capture/mxc_v4l2_capture.c
index 4b1f49c..6ddfbdb 100644
--- a/drivers/media/video/mxc/capture/mxc_v4l2_capture.c
+++ b/drivers/media/video/mxc/capture/mxc_v4l2_capture.c
@@ -98,18 +98,18 @@ static video_fmt_t video_fmts[] = {
 	{			/*! NTSC */
 	 .v4l2_id = V4L2_STD_NTSC,
 	 .name = "NTSC",
-	 .raw_width = 720,		/* SENS_FRM_WIDTH */
-	 .raw_height = 288,		/* SENS_FRM_HEIGHT */
-	 .active_width = 720,		/* ACT_FRM_WIDTH */
-	 .active_height = (480 / 2),	/* ACT_FRM_HEIGHT */
+	 .raw_width = 720 - 1,		/* SENS_FRM_WIDTH */
+	 .raw_height = 288 - 1,		/* SENS_FRM_HEIGHT */
+	 .active_width = 720,		/* ACT_FRM_WIDTH plus 1 */
+	 .active_height = (480 / 2),	/* ACT_FRM_HEIGHT plus 1 */
 	 .active_top = 12,
 	 .active_left = 0,
 	 },
 	{			/*! (B, G, H, I, N) PAL */
 	 .v4l2_id = V4L2_STD_PAL,
 	 .name = "PAL",
-	 .raw_width = 720,
-	 .raw_height = (576 / 2) + 24 * 2,
+	 .raw_width = 720 - 1,
+	 .raw_height = (576 / 2) + 24 * 2 - 1,
 	 .active_width = 720,
 	 .active_height = (576 / 2),
 	 .active_top = 0,
@@ -118,8 +118,8 @@ static video_fmt_t video_fmts[] = {
 	{			/*! Unlocked standard */
 	 .v4l2_id = V4L2_STD_ALL,
 	 .name = "Autodetect",
-	 .raw_width = 720,
-	 .raw_height = (576 / 2) + 24 * 2,
+	 .raw_width = 720 - 1,
+	 .raw_height = (576 / 2) + 24 * 2 - 1,
 	 .active_width = 720,
 	 .active_height = (576 / 2),
 	 .active_top = 0,
@@ -2265,12 +2265,12 @@ static int mxc_v4l2_resume(struct platform_device *pdev)
 
 	cam->low_power = false;
 	wake_up_interruptible(&cam->power_queue);
+	camera_power(cam, true);
 
 	if (cam->overlay_on == true)
 		start_preview(cam);
 	if (cam->capture_on == true)
 		mxc_streamon(cam);
-	camera_power(cam, true);
 
 	return 0;
 }
diff --git a/drivers/mxc/ipu/ipu_common.c b/drivers/mxc/ipu/ipu_common.c
index 13e7493..ca3bf85 100644
--- a/drivers/mxc/ipu/ipu_common.c
+++ b/drivers/mxc/ipu/ipu_common.c
@@ -1765,7 +1765,13 @@ static int ipu_suspend(struct platform_device *pdev, pm_message_t state)
 			g_channel_init_mask_backup = g_channel_init_mask;
 			g_channel_init_mask |= 2;
 		}
+	} else if (cpu_is_mx35()) {
+		g_ipu_config = __raw_readl(IPU_CONF);
+		/* Disable the clock of display otherwise the backlight cannot
+		 * be close after camera/tvin related test */
+		__raw_writel(g_ipu_config & 0xfbf, IPU_CONF);
 	}
+
 	return 0;
 }
 
@@ -1778,7 +1784,10 @@ static int ipu_resume(struct platform_device *pdev)
 			clk_disable(g_ipu_csi_clk);
 			g_channel_init_mask = g_channel_init_mask_backup;
 		}
-	}
+	} else if (cpu_is_mx35()) {
+		__raw_writel(g_ipu_config, IPU_CONF);
+    }
+
 	return 0;
 }
 
-- 
1.5.4.4

